src/btree/bt_walk.c

```diff
@@ -259,12 +261,17 @@ __tree_walk_internal(WT_SESSION_IMPL *session, WT_REF **refp, uint64_t *walkcntp
     WT_ASSERT(session, LF_ISSET(WT_READ_VISIBLE_ALL) || F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT));
 
     /*
-     * All tree walks except for rollback-to-stable skip deleted pages. We set read-skip-deleted
-     * here because we didn't want to add a flag to all of the tree-walk callers, and we make it
-     * worse because we don't want to add a flag that turns the read-skip-deleted flag off, so we
-     * test the RTS flag itself.
+     * Historically, all tree walks skipped deleted pages. There are now some exceptions to this:
+     * Rollback to stable, and column store append. Rather than add the read-see-deleted flag to
+     * every single tree walk call, we hide these pages unless:
+     *
+     * 1. We detect that rollback to stable is in progress
+     * 2. Callers opt into seeing these pages with the read-see-deleted flag.
+     *
+     * Ideally, rollback to stable would also use the read-see-deleted flag but it uses cursor->next
+     * and cursor->prev, which don't have flags.
      */
-    if (!F_ISSET(session, WT_SESSION_ROLLBACK_TO_STABLE))
+    if (!F_ISSET(session, WT_SESSION_ROLLBACK_TO_STABLE) && !LF_ISSET(WT_READ_SEE_DELETED))
```

Comment:
Could this be simpified to no longer explicitly check for `WT_SESSION_ROLLBACK_TO_STABLE`? I like the idea of using a flag passed instead - that has nicer layering characteristics.

Reply:
I had a go at this initially, I also preferred the idea of using a flag. Unfortunately RTS can end up in `__tree_walk_internal` due to `cursor->next` and `prev`, which don't accept flags. 

So it'd be a larger refactor, involving those cursor functions somehow remembering that RTS is doing the walk (cursor open config?) and applying that to the `__tree_walk_internal` call.

Reply:
RTS shouldn't call any `cursor->next` or `prev` for the data store pages. It is possible with the `history store` cursor. There is no problem in skipping the WT_REF_DELETED history store pages. History store pages are marked as WT_REF_DELETED only when everything on the page is obsolete. Removing the RTS session check shouldn't cause any problem.

Reply:
I'm interested in understanding this. It doesn't make logical sense to me that both are needed at the moment.
OTOH: refactoring the walk code is also something I'd prefer to avoid right now, so I won't push hard.

Reply:
> RTS shouldn't call any cursor->next or prev for the data store pages.

Maybe that's how it works right now - but I think it's a real footgun that a single `cursor->next` or `prev` in RTS could blow up.

If we're not going to change `cursor->next`/`prev` to take flags (and I don't think we should), then the real fix here (IMO) would for `__wt_btcur_open` to take a `flags` argument and for that to change the behaviour of the cursor.

The way the code is designed at the moment, it forces you to either have a hacky session flags check for RTS, or accept that `cursor->next` or `prev` will blow up when called from RTS.

Reply:
Hari and I discussed this offline. We should revisit the RTS session flags thing in WT-11683. I think that will naturally happen since that code sets the `WT_READ_VISIBLE_ALL` flag.